In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings

# 경고 메시지 출력 표기 생략
warnings.filterwarnings('ignore')

## 1. 데이터 불러오기

▶ CSV파일 데이터 불러오기
> **pd.read_csv (파일경로), pd.read_excel (파일경로)** <br>
> 이때 파일경로는 문자열 형태로 입력
>
> 예) pd.read_csv("/content/sample_data/sample.csv") <br>

In [2]:
# 데이터 불러오기
# 파일 경로 = "https://raw.githubusercontent.com/agtechresearch/LectureAlgorithm/main/csv/married_simp.csv"
married_dataset = pd.read_csv("https://raw.githubusercontent.com/agtechresearch/LectureAlgorithm/main/csv/married_simp.csv")
married_dataset

gender   age  age_partner  importance_same_religion  \
0     female  21.0         27.0                       4.0   
1     female  21.0         22.0                       4.0   
2     female  21.0         22.0                       4.0   
3     female  21.0         23.0                       4.0   
4     female  21.0         24.0                       4.0   
...      ...   ...          ...                       ...   
8373    male  25.0         26.0                       1.0   
8374    male  25.0         24.0                       1.0   
8375    male  25.0         29.0                       1.0   
8376    male  25.0         22.0                       1.0   
8377    male  25.0         22.0                       1.0   

      pref_of_partner_attractive  pref_of_partner_sincere  \
0                           35.0                     20.0   
1                           60.0                      0.0   
2                           19.0                     18.0   
3                           30.0                      5.0   
4                           30.0                     10.0   
...                          ...                      ...   
8373                        10.0                     10.0   
8374                        50.0                     20.0   
8375                        40.0                     10.0   
8376                        10.0                     25.0   
8377                        20.0                     20.0   

      pref_of_partner_funny  pref_of_partner_shared_interests  \
0                      20.0                               5.0   
1                      40.0                               0.0   
2                      18.0                              12.0   
3                      40.0                               5.0   
4                      10.0                              20.0   
...                     ...                               ...   
8373                   20.0                              15.0   
8374                    5.0                               5.0   
8375                   10.0                               NaN   
8376                   10.0                              20.0   
8377                   15.0                              30.0   

      eval_by_partner_attractive  eval_by_partner_sincere  ...  my_pref_funny  \
0                            6.0                      8.0  ...           15.0   
1                            7.0                      8.0  ...           15.0   
2                           10.0                     10.0  ...           15.0   
3                            7.0                      8.0  ...           15.0   
4                            8.0                      7.0  ...           15.0   
...                          ...                      ...  ...            ...   
8373                        10.0                      5.0  ...           15.0   
8374                         6.0                      3.0  ...           15.0   
8375                         2.0                      1.0  ...           15.0   
8376                         5.0                      7.0  ...           15.0   
8377                         8.0                      8.0  ...           15.0   

      my_pref_shared_interests  my_eval_attractive  my_eval_sincere  \
0                         15.0                 6.0              9.0   
1                         15.0                 7.0              8.0   
2                         15.0                 5.0              8.0   
3                         15.0                 7.0              6.0   
4                         15.0                 5.0              6.0   
...                        ...                 ...              ...   
8373                       0.0                 3.0              5.0   
8374                       0.0                 4.0              6.0   
8375                       0.0                 4.0              7.0   
8376                       0.0                 4.0              6.0   
8377 

In [3]:
# 변수가 24개가 있으므로, 모든 변수를 출력하려면 다음과 같이 설정해야 함
pd.options.display.max_columns = 25
married_dataset

gender   age  age_partner  importance_same_religion  \
0     female  21.0         27.0                       4.0   
1     female  21.0         22.0                       4.0   
2     female  21.0         22.0                       4.0   
3     female  21.0         23.0                       4.0   
4     female  21.0         24.0                       4.0   
...      ...   ...          ...                       ...   
8373    male  25.0         26.0                       1.0   
8374    male  25.0         24.0                       1.0   
8375    male  25.0         29.0                       1.0   
8376    male  25.0         22.0                       1.0   
8377    male  25.0         22.0                       1.0   

      pref_of_partner_attractive  pref_of_partner_sincere  \
0                           35.0                     20.0   
1                           60.0                      0.0   
2                           19.0                     18.0   
3                           30.0                      5.0   
4                           30.0                     10.0   
...                          ...                      ...   
8373                        10.0                     10.0   
8374                        50.0                     20.0   
8375                        40.0                     10.0   
8376                        10.0                     25.0   
8377                        20.0                     20.0   

      pref_of_partner_funny  pref_of_partner_shared_interests  \
0                      20.0                               5.0   
1                      40.0                               0.0   
2                      18.0                              12.0   
3                      40.0                               5.0   
4                      10.0                              20.0   
...                     ...                               ...   
8373                   20.0                              15.0   
8374                    5.0                               5.0   
8375                   10.0                               NaN   
8376                   10.0                              20.0   
8377                   15.0                              30.0   

      eval_by_partner_attractive  eval_by_partner_sincere  \
0                            6.0                      8.0   
1                            7.0                      8.0   
2                           10.0                     10.0   
3                            7.0                      8.0   
4                            8.0                      7.0   
...                          ...                      ...   
8373                        10.0                      5.0   
8374                         6.0                      3.0   
8375                         2.0                      1.0   
8376                         5.0                      7.0   
8377                         8.0                      8.0   

      eval_by_partner_funny  eval_by_partner_shared_interests  \
0                       8.0                               6.0   
1                       7.0                               5.0   
2                      10.0                              10.0   
3                       8.0                               8.0   
4                       6.0                               7.0   
...                     ...                               ...   
8373                    2.0                               5.0   
8374                    3.0                               2.0   
8375                    2.0                               1.0   
8376                    5.0                               6.0   
8377                    7.0                               7.0   

      my_pref_attractive  my_pref_sincere  my_pref_funny  \
0                   15.0             20.0           15.0   
1                   15.0             20.0           15.0   
2                   15.0             20.0           15.0   
3    

▶ 결혼정보업체 데이터셋

- **gender**: 성별(female, male)

- **age**: 나이 (my age)

- **age_partner**: 상대방 나이  (partner's age)

- **importance_same_religion**: 종교를 중요시 여기는 지에 대한 응답 (1-10)

- **pref_of_partner_attractive**: 매력(attractive)에 대하여 상대방이 얼마나 중요하게 생각하는지의 여부

- **pref_of_partner_sincere**: 진정성/성실성(sincere)에 대하여 상대방이 얼마나 중요하게 생각하는지의 여부

- **pref_of_partner_funny**: 유머감각(funny)에 대하여 상대방이 얼마나 중요하게 생각하는지의 여부

- **pref_of_partner_shared_interests**: 공통의 관심사(shared interests)에 대하여 상대방이 얼마나 중요하게 생각하는지의 여부

- **eval_by_partner_attractive**: 상대방이 나의 매력(attractive)에 대하여 평가한 값

- **eval_by_partner_sincere**: 상대방이 나의 진정성/성실성(sincere)에 대하여 평가한 값

- **eval_by_partner_funny**: 상대방이 나의 유머감각(funny)에 대하여 평가한 값

- **eval_by_partner_shared_interests**: 상대방이 나와 얼마만큼의 공통 관심사(shared interests)가 있는지를 평가한 값

- **my_pref_attractive**: 매력(attractive)에 대하여 내가 얼마나 중요하게 생각하는지의 여부

- **my_pref_sincere**: 진정성/성실성(sincere)에 대하여 내가 얼마나 중요하게 생각하는지의 여부

- **my_pref_funny**: 유머감각에 대하여 내가 얼마나 중요하게 생각하는지의 여부

- **my_pref_shared_interests**: 공통 관심사에 대하여 내가 얼마나 중요하게 생각하는지의 여부

- **my_eval_attractive**: 내가 상대방의 매력에 대하여 평가한 값

- **my_eval_sincere**: 내가 상대방의 진정성/성실성에 대하여 평가한 값

- **my_eval_funny**: 내가 상대방의 유머감각에 대하여 평가한 값

- **my_eval_shared_interests**: 내가 상대방과의 얼마만큼의 공통 관심사가 있는지를 평가한 값

- **interests_correlate**: 서로간의 공통 관심사 연관도

- **how_much_i_liked**: 내가 상대방이 얼마나 마음에 들었는지의 여부(값)

- **guess_prob_liked**: 상대방이 나를 얼마나 마음에 들어하였을지의 여부(예상값)

- **married**: 만남 이후 결혼 성사 여부

## 2. 데이터셋 확인
▶ 전반적인 데이터셋 정보 확인 (**.info()**) <br>
> 데이터셋의 **컬럼정보와 값의 수, 각 컬럼별 데이터타입, 행의 갯수 등**의 정보 <br>

In [4]:
# 데이터셋 정보 확인
married_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8378 entries, 0 to 8377
Data columns (total 24 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   gender                            8378 non-null   object 
 1   age                               8283 non-null   float64
 2   age_partner                       8274 non-null   float64
 3   importance_same_religion          8299 non-null   float64
 4   pref_of_partner_attractive        8289 non-null   float64
 5   pref_of_partner_sincere           8289 non-null   float64
 6   pref_of_partner_funny             8280 non-null   float64
 7   pref_of_partner_shared_interests  8249 non-null   float64
 8   eval_by_partner_attractive        8166 non-null   float64
 9   eval_by_partner_sincere           8091 non-null   float64
 10  eval_by_partner_funny             8018 non-null   float64
 11  eval_by_partner_shared_interests  7302 non-null   float64
 12  my_pre

In [5]:
# 데이터의 통계량 확인하기
married_dataset.describe()

age  age_partner  importance_same_religion  \
count  8283.000000  8274.000000               8299.000000   
mean     26.358928    26.364999                  3.651645   
std       3.566763     3.563648                  2.805237   
min      18.000000    18.000000                  1.000000   
25%      24.000000    24.000000                  1.000000   
50%      26.000000    26.000000                  3.000000   
75%      28.000000    28.000000                  6.000000   
max      55.000000    55.000000                 10.000000   

       pref_of_partner_attractive  pref_of_partner_sincere  \
count                 8289.000000              8289.000000   
mean                    22.495347                17.396867   
std                     12.569802                 7.044003   
min                      0.000000                 0.000000   
25%                     15.000000                15.000000   
50%                     20.000000                18.370000   
75%                     25.000000                20.000000   
max                    100.000000                60.000000   

       pref_of_partner_funny  pref_of_partner_shared_interests  \
count            8280.000000                       8249.000000   
mean               17.459714                         11.845930   
std                 6.085526                          6.362746   
min                 0.000000                          0.000000   
25%                15.000000                          9.520000   
50%                18.000000                         10.640000   
75%                20.000000                         16.000000   
max                50.000000                         30.000000   

       eval_by_partner_attractive  eval_by_partner_sincere  \
count                 8166.000000              8091.000000   
mean                     6.190411                 7.175256   
std                      1.950305                 1.740575   
min                      0.000000                 0.000000   
25%                      5.000000                 6.000000   
50%                      6.000000                 7.000000   
75%                      8.000000                 8.000000   
max                     10.500000                10.000000   

       eval_by_partner_funny  eval_by_partner_shared_interests  \
count            8018.000000                       7302.000000   
mean                6.400599                          5.474870   
std                 1.954078                          2.156163   
min                 0.000000                          0.000000   
25%                 5.000000                          4.000000   
50%                 7.000000                          6.000000   
75%                 8.000000                          7.000000   
max                11.000000                         10.000000   

       my_pref_attractive  my_pref_sincere  my_pref_funny  \
count         8299.000000      8299.000000    8289.000000   
mean            22.514632        17.396389      17.457043   
std             12.587674         7.046700       6.085239   
min              0.000000         0.000000       0.000000   
25%             15.000000        15.000000      15.000000   
50%             20.000000        18.180000      18.000000   
75%             25.000000        20.000000      20.000000   
max            100.000000        60.000000      50.000000   

       my_pref_shared_interests  my_eval_attractive  my_eval_sincere  \
count               8257.000000         8176.000000      8101.000000   
mean                  11.845111            6.189995         7.175164   
std                    6.362154            1.950169         1.740315   
min                    0.000000            0.000000         0.000000   
25%                    9.520000            5.000000         6.000000   
50%                   10.640000            6.000000         7.000000   
75%                   16.000000            8.000000         8.000000   
max                   30.000000

In [6]:
round(married_dataset.describe(), 2)

age  age_partner  importance_same_religion  \
count  8283.00      8274.00                   8299.00   
mean     26.36        26.36                      3.65   
std       3.57         3.56                      2.81   
min      18.00        18.00                      1.00   
25%      24.00        24.00                      1.00   
50%      26.00        26.00                      3.00   
75%      28.00        28.00                      6.00   
max      55.00        55.00                     10.00   

       pref_of_partner_attractive  pref_of_partner_sincere  \
count                     8289.00                  8289.00   
mean                        22.50                    17.40   
std                         12.57                     7.04   
min                          0.00                     0.00   
25%                         15.00                    15.00   
50%                         20.00                    18.37   
75%                         25.00                    20.00   
max                        100.00                    60.00   

       pref_of_partner_funny  pref_of_partner_shared_interests  \
count                8280.00                           8249.00   
mean                   17.46                             11.85   
std                     6.09                              6.36   
min                     0.00                              0.00   
25%                    15.00                              9.52   
50%                    18.00                             10.64   
75%                    20.00                             16.00   
max                    50.00                             30.00   

       eval_by_partner_attractive  eval_by_partner_sincere  \
count                     8166.00                  8091.00   
mean                         6.19                     7.18   
std                          1.95                     1.74   
min                          0.00                     0.00   
25%                          5.00                     6.00   
50%                          6.00                     7.00   
75%                          8.00                     8.00   
max                         10.50                    10.00   

       eval_by_partner_funny  eval_by_partner_shared_interests  \
count                8018.00                           7302.00   
mean                    6.40                              5.47   
std                     1.95                              2.16   
min                     0.00                              0.00   
25%                     5.00                              4.00   
50%                     7.00                              6.00   
75%                     8.00                              7.00   
max                    11.00                             10.00   

       my_pref_attractive  my_pref_sincere  my_pref_funny  \
count             8299.00          8299.00        8289.00   
mean                22.51            17.40          17.46   
std                 12.59             7.05           6.09   
min                  0.00             0.00           0.00   
25%                 15.00            15.00          15.00   
50%                 20.00            18.18          18.00   
75%                 25.00            20.00          20.00   
max                100.00            60.00          50.00   

       my_pref_shared_interests  my_eval_attractive  my_eval_sincere  \
count                   8257.00             8176.00          8101.00   
mean                      11.85                6.19             7.18   
std                        6.36                1.95             1.74   
min                        0.00                0.00             0.00   
25%                        9.52                5.00             6.00   
50%                       10.64                6.00             7.00   
75%                       16.00                8.00             8.00   
max                       30.00               10.00            

▶ 최댓값을 살펴보면 다소 특이한 부분이 있음

- 본인 및 상대방을 평가하는 변수들(eval_by_partner_xxx, my_eval_xxx)에서는 대체로 최댓값이 10 으로 일정함

- 중요도와 관련된 변수(pref_of_partner_xxx, my_pref_xxx)에서는 최댓값이 50, 60, 100 등으로 다양함

## 3. 전처리: (1) 결측치 처리

In [7]:
# 항목별 결측치 비율 확인
married_dataset.isna().mean()

gender                              0.000000
age                                 0.011339
age_partner                         0.012413
importance_same_religion            0.009429
pref_of_partner_attractive          0.010623
pref_of_partner_sincere             0.010623
pref_of_partner_funny               0.011697
pref_of_partner_shared_interests    0.015397
eval_by_partner_attractive          0.025304
eval_by_partner_sincere             0.034256
eval_by_partner_funny               0.042970
eval_by_partner_shared_interests    0.128432
my_pref_attractive                  0.009429
my_pref_sincere                     0.009429
my_pref_funny                       0.010623
my_pref_shared_interests            0.014443
my_eval_attractive                  0.024111
my_eval_sincere                     0.033063
my_eval_funny                       0.041776
my_eval_shared_interests            0.127357
interests_correlate                 0.018859
how_much_i_liked                    0.028646
guess_prob

In [8]:
# 결측치 처리
married_dataset = married_dataset.dropna(axis=0)
married_dataset

gender   age  age_partner  importance_same_religion  \
0     female  21.0         27.0                       4.0   
1     female  21.0         22.0                       4.0   
3     female  21.0         23.0                       4.0   
4     female  21.0         24.0                       4.0   
5     female  21.0         25.0                       4.0   
...      ...   ...          ...                       ...   
8368    male  25.0         26.0                       1.0   
8369    male  25.0         22.0                       1.0   
8370    male  25.0         27.0                       1.0   
8376    male  25.0         22.0                       1.0   
8377    male  25.0         22.0                       1.0   

      pref_of_partner_attractive  pref_of_partner_sincere  \
0                           35.0                     20.0   
1                           60.0                      0.0   
3                           30.0                      5.0   
4                           30.0                     10.0   
5                           50.0                      0.0   
...                          ...                      ...   
8368                        10.0                     30.0   
8369                        10.0                     20.0   
8370                        10.0                     25.0   
8376                        10.0                     25.0   
8377                        20.0                     20.0   

      pref_of_partner_funny  pref_of_partner_shared_interests  \
0                      20.0                               5.0   
1                      40.0                               0.0   
3                      40.0                               5.0   
4                      10.0                              20.0   
5                      10.0                              10.0   
...                     ...                               ...   
8368                   15.0                              10.0   
8369                   20.0                              20.0   
8370                   20.0                              20.0   
8376                   10.0                              20.0   
8377                   15.0                              30.0   

      eval_by_partner_attractive  eval_by_partner_sincere  \
0                            6.0                      8.0   
1                            7.0                      8.0   
3                            7.0                      8.0   
4                            8.0                      7.0   
5                            7.0                      7.0   
...                          ...                      ...   
8368                         9.0                      7.0   
8369                         8.0                      9.0   
8370                         7.0                      2.0   
8376                         5.0                      7.0   
8377                         8.0                      8.0   

      eval_by_partner_funny  eval_by_partner_shared_interests  \
0                       8.0                               6.0   
1                       7.0                               5.0   
3                       8.0                               8.0   
4                       6.0                               7.0   
5                       8.0                               7.0   
...                     ...                               ...   
8368                    8.0                               6.0   
8369                    7.0                               7.0   
8370                    5.0                               3.0   
8376                    5.0                               6.0   
8377                    7.0                               7.0   

      my_pref_attractive  my_pref_sincere  my_pref_funny  \
0                   15.0             20.0           15.0   
1                   15.0             20.0           15.0   
3                   15.0             20.0           15.0   
4    

In [9]:
# object 타입의 문자열 변수를 숫자형으로 변환
married_dataset = pd.get_dummies(married_dataset, columns=['gender'], drop_first=True)

In [10]:
married_dataset.head(5)

age  age_partner  importance_same_religion  pref_of_partner_attractive  \
0  21.0         27.0                       4.0                        35.0   
1  21.0         22.0                       4.0                        60.0   
3  21.0         23.0                       4.0                        30.0   
4  21.0         24.0                       4.0                        30.0   
5  21.0         25.0                       4.0                        50.0   

   pref_of_partner_sincere  pref_of_partner_funny  \
0                     20.0                   20.0   
1                      0.0                   40.0   
3                      5.0                   40.0   
4                     10.0                   10.0   
5                      0.0                   10.0   

   pref_of_partner_shared_interests  eval_by_partner_attractive  \
0                               5.0                         6.0   
1                               0.0                         7.0   
3                               5.0                         7.0   
4                              20.0                         8.0   
5                              10.0                         7.0   

   eval_by_partner_sincere  eval_by_partner_funny  \
0                      8.0                    8.0   
1                      8.0                    7.0   
3                      8.0                    8.0   
4                      7.0                    6.0   
5                      7.0                    8.0   

   eval_by_partner_shared_interests  my_pref_attractive  my_pref_sincere  \
0                               6.0                15.0             20.0   
1                               5.0                15.0             20.0   
3                               8.0                15.0             20.0   
4                               7.0                15.0             20.0   
5                               7.0                15.0             20.0   

   my_pref_funny  my_pref_shared_interests  my_eval_attractive  \
0           15.0                      15.0                 6.0   
1           15.0                      15.0                 7.0   
3           15.0                      15.0                 7.0   
4           15.0                      15.0                 5.0   
5           15.0                      15.0                 4.0   

   my_eval_sincere  my_eval_funny  my_eval_shared_interests  \
0              9.0            7.0                       5.0   
1              8.0            8.0                       6.0   
3              6.0            7.0                       8.0   
4              6.0            7.0                       6.0   
5              9.0            4.0                       4.0   

   interests_correlate  how_much_i_liked  guess_prob_liked  married  \
0                 0.14               7.0               6.0        0   
1                 0.54               7.0               5.0        0   
3                 0.61               7.0               6.0        1   
4                 0.21               6.0               6.0        1   
5                 0.25               6.0               5.0        0   

   gender_male  
0        False  
1        False  
3        False  
4        False  
5        False

## 4. 모델 학습 및 평가


In [11]:
# 데이터셋, 독립변수와 종속변수 분리: 독립변수 -> x, 종속변수 -> y
x = married_dataset.drop(['married'], axis=1)
y = married_dataset['married']

In [12]:
x

age  age_partner  importance_same_religion  pref_of_partner_attractive  \
0     21.0         27.0                       4.0                        35.0   
1     21.0         22.0                       4.0                        60.0   
3     21.0         23.0                       4.0                        30.0   
4     21.0         24.0                       4.0                        30.0   
5     21.0         25.0                       4.0                        50.0   
...    ...          ...                       ...                         ...   
8368  25.0         26.0                       1.0                        10.0   
8369  25.0         22.0                       1.0                        10.0   
8370  25.0         27.0                       1.0                        10.0   
8376  25.0         22.0                       1.0                        10.0   
8377  25.0         22.0                       1.0                        20.0   

      pref_of_partner_sincere  pref_of_partner_funny  \
0                        20.0                   20.0   
1                         0.0                   40.0   
3                         5.0                   40.0   
4                        10.0                   10.0   
5                         0.0                   10.0   
...                       ...                    ...   
8368                     30.0                   15.0   
8369                     20.0                   20.0   
8370                     25.0                   20.0   
8376                     25.0                   10.0   
8377                     20.0                   15.0   

      pref_of_partner_shared_interests  eval_by_partner_attractive  \
0                                  5.0                         6.0   
1                                  0.0                         7.0   
3                                  5.0                         7.0   
4                                 20.0                         8.0   
5                                 10.0                         7.0   
...                                ...                         ...   
8368                              10.0                         9.0   
8369                              20.0                         8.0   
8370                              20.0                         7.0   
8376                              20.0                         5.0   
8377                              30.0                         8.0   

      eval_by_partner_sincere  eval_by_partner_funny  \
0                         8.0                    8.0   
1                         8.0                    7.0   
3                         8.0                    8.0   
4                         7.0                    6.0   
5                         7.0                    8.0   
...                       ...                    ...   
8368                      7.0                    8.0   
8369                      9.0                    7.0   
8370                      2.0                    5.0   
8376                      7.0                    5.0   
8377                      8.0                    7.0   

      eval_by_partner_shared_interests  my_pref_attractive  my_pref_sincere  \
0                                  6.0                15.0             20.0   
1                                  5.0                15.0             20.0   
3                                  8.0                15.0             20.0   
4                                  7.0                15.0             20.0   
5                                  7.0                15.0             20.0   
...                                ...                 ...              ...   
8368                               6.0                70.0              0.0   
8369                               7.0                70.0              0.0   
8370                               3.0                70.0              0.0   
8376                               6.0                70.

In [13]:
y

0       0
1       0
3       1
4       1
5       0
       ..
8368    0
8369    0
8370    0
8376    0
8377    0
Name: married, Length: 6055, dtype: int64

In [14]:
# train set 과 test set 으로 데이터를 나누기 위해 train_test_split 함수를 불러옴
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state=100)

In [15]:
import xgboost as xgb

#model = xgb.XGBClassifier(n_estimators=500, max_depth=5, random_state=100)
model = xgb.XGBClassifier(random_state=100)
model.fit(X_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              num_parallel_tree=None, random_state=100, ...)

In [16]:
pred = model.predict(X_test)  # 테스트 데이터로 예측
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
accuracy_score(y_test, pred)  # 정확도

0.8406275805119736

In [17]:
print(confusion_matrix(y_test, pred))  # confusion matrix 출력

[[935  63]
 [130  83]]


- 실제값이 0 인데 0 으로 예측한 경우가 935 건으로 가장 많음

- 결혼에 성공한 경우를 제대로 예측한 경우도 83 건으로 무난한 수준임

- 1종 오류, 즉 실제값은 0 인데 1 로 예측한 경우가 63 건

- 2종 오류, 즉 실제값은 1 인데 0 으로 예측한 경우가 130 건

In [18]:
print(classification_report(y_test, pred))  # classification report 출력

              precision    recall  f1-score   support

           0       0.88      0.94      0.91       998
           1       0.57      0.39      0.46       213

    accuracy                           0.84      1211
   macro avg       0.72      0.66      0.68      1211
weighted avg       0.82      0.84      0.83      1211



## 5. 하이퍼파라미터 튜닝 (Hyper-parameter tuning): Grid search 

In [19]:
from sklearn.model_selection import GridSearchCV

In [21]:
# 하이퍼파라미터 튜닝을 위한 후보 파라미터들의 범위(set) 설정
parameters = {
    'learning_rate': [0.1, 0.2, 0.3],  # 학습률
    'max_depth': [2, 3, 5],             # 각 트리의 최대 깊이 제한
    'subsample': [0.5, 0.7, 0.8],         # 각 트리마다 사용할 훈련 데이터의 비율 (예 0.5를 쓰면 데이터의 절반씩만 랜덤 추출하여 트리 생성) -> 오버피팅 방지에 사용
    'n_estimators': [100, 300, 500]    # 전체 트리의 개수
}

In [22]:
# CPU 코어 8개를 사용하여 학습
model_tuning = xgb.XGBClassifier(random_state=100)
gridsearch_model = GridSearchCV(model_tuning, parameters, scoring='f1', cv=5, n_jobs=8, verbose=1)
# scoring='f1' -> f1 score 로 평가
# cv=5 -> 5-fold cross validation
# verbose=1 -> 학습 과정 출력

## GPU 사용 시
#model_tuning = xgb.XGBClassifier(tree_method='gpu_hist', gpu_id=0)
#gridsearch_model = GridSearchCV(model_tuning, parameters, scoring='f1', cv=5, verbose=1)

In [23]:
gridsearch_model.fit(X_train, y_train)

Fitting 5 folds for each of 81 candidates, totalling 405 fits


GridSearchCV(cv=5,
             estimator=XGBClassifier(base_score=None, booster=None,
                                     callbacks=None, colsample_bylevel=None,
                                     colsample_bynode=None,
                                     colsample_bytree=None, device=None,
                                     early_stopping_rounds=None,
                                     enable_categorical=False, eval_metric=None,
                                     feature_types=None, gamma=None,
                                     grow_policy=None, importance_type=None,
                                     interaction_constraints=None,
                                     learning_rate=None,...
                                     max_delta_step=None, max_depth=None,
                                     max_leaves=None, min_child_weight=None,
                                     missing=nan, monotone_constraints=None,
                                     multi_strategy=None, n_estimators=None,
                                     n_jobs=None, num_parallel_tree=None,
                                     random_state=100, ...),
             n_jobs=8,
             param_grid={'learning_rate': [0.1, 0.2, 0.3],
                         'max_depth': [2, 3, 5],
                         'n_estimators': [100, 300, 500],
                         'subsample': [0.5, 0.7, 0.8]},
             scoring='f1', verbose=1)

* 학습 완료까지 꽤 많은 시간이 소요됨
* 하이퍼파라미터 4종류 당 3개씩: 총 81번 모델링 진행(3^4) + 교차검증 5회 진행 => 총 405번 모델링 작업 수행

In [24]:
gridsearch_model.best_params_

{'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 500, 'subsample': 0.7}

In [25]:
pred_tuning = gridsearch_model.predict(X_test)  # 테스트 데이터로 예측
accuracy_score(y_test, pred_tuning)  # 정확도

0.8422791081750619